pkg/protocol/engine/congestioncontrol/scheduler/drr/validatorqueue.go

```diff
@@ -0,0 +1,169 @@
+package drr
+
+import (
+	"container/heap"
+	"fmt"
+	"time"
+
+	"go.uber.org/atomic"
+
+	"github.com/iotaledger/hive.go/ds/generalheap"
+	"github.com/iotaledger/hive.go/ds/shrinkingmap"
+	"github.com/iotaledger/hive.go/lo"
+	"github.com/iotaledger/hive.go/runtime/timed"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/blocks"
+
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+type ValidatorQueue struct {
+	accountID iotago.AccountID
+	submitted *shrinkingmap.ShrinkingMap[iotago.BlockID, *blocks.Block]
+	inbox     generalheap.Heap[timed.HeapKey, *blocks.Block]
+	size      atomic.Int64
+
+	tokenBucket      float64
+	lastScheduleTime time.Time
+
+	blockChan      chan *blocks.Block
+	shutdownSignal chan struct{}
+}
+
+func NewValidatorQueue(accountID iotago.AccountID) *ValidatorQueue {
+	return &ValidatorQueue{
+		accountID:      accountID,
+		submitted:      shrinkingmap.New[iotago.BlockID, *blocks.Block](),
+		blockChan:      make(chan *blocks.Block, 1),
+		shutdownSignal: make(chan struct{}),
+	}
+}
+
+func (q *ValidatorQueue) Size() int {
+	if q == nil {
+		return 0
+	}
+
+	return int(q.size.Load())
+}
+
+func (q *ValidatorQueue) AccountID() iotago.AccountID {
+	return q.accountID
+}
+
+func (q *ValidatorQueue) Submit(block *blocks.Block) bool {
+	if blkAccountID := block.ProtocolBlock().IssuerID; q.accountID != blkAccountID {
+		panic(fmt.Sprintf("issuerqueue: queue issuer ID(%x) and issuer ID(%x) does not match.", q.accountID, blkAccountID))
+	}
+
+	if _, submitted := q.submitted.Get(block.ID()); submitted {
+		return false
+	}
+
+	q.submitted.Set(block.ID(), block)
+	q.size.Inc()
+
+	return true
+}
+
+func (q *ValidatorQueue) Unsubmit(block *blocks.Block) bool {
+	if _, submitted := q.submitted.Get(block.ID()); !submitted {
+		return false
+	}
+
+	q.submitted.Delete(block.ID())
+	q.size.Dec()
+
+	return true
+}
+
+func (q *ValidatorQueue) Ready(block *blocks.Block) bool {
+	if _, submitted := q.submitted.Get(block.ID()); !submitted {
+		return false
+	}
+
+	q.submitted.Delete(block.ID())
+	heap.Push(&q.inbox, &generalheap.HeapElement[timed.HeapKey, *blocks.Block]{Value: block, Key: timed.HeapKey(block.IssuingTime())})
+
+	return true
+}
+
+// PopFront removes the first ready block from the queue.
+func (q *ValidatorQueue) PopFront() *blocks.Block {
+	if q.inbox.Len() == 0 {
+		return nil
+	}
+
+	heapElement, isHeapElement := heap.Pop(&q.inbox).(*generalheap.HeapElement[timed.HeapKey, *blocks.Block])
+	if !isHeapElement {
+		return nil
+	}
+	blk := heapElement.Value
+	q.size.Dec()
+
+	return blk
+}
+
+func (q *ValidatorQueue) RemoveTail() *blocks.Block {
```

Comment:
This method is not used. I think ValidatorQueue should have a limited size and the tail should be removed if that is exceeded, just like for regular blocks.